exodus_gw/routers/dynamodb.py

```diff
@@ -0,0 +1,113 @@
+import logging
+from typing import List, Tuple
+
+import backoff
+
+from ..aws.client import DynamoDBClientWrapper as ddb_client
+from ..settings import get_environment
+
+LOG = logging.getLogger("exodus-gw")
+
+
+@backoff.on_predicate(
+    backoff.expo,
+    lambda response: response["UnprocessedItems"],
+    max_time=600,
+)
+async def _batch_write_rollback(details: dict):
+    """Deletes all items written by the failed batch_write operation.
+    Triggered by the 'backoff.on_predicate' decorator.
+
+    Retries automatically, with exponential backoff and jitter, if
+    there are any unprocessed items.
+
+    'details' dictionary, provided by the backoff decorator, includes:
+    - target: reference to the function or method being invoked
+    - args: positional arguments to func
+    - kwargs: keyword arguments to func
+    - tries: number of invocation tries so far
+    - elapsed: elapsed time in seconds so far
+    """
+
+    env = details["args"][0]
+    items = details["args"][1]
+
+    env_obj = get_environment(env)
+    profile = env_obj.aws_profile
+    table = env_obj.table
+
+    try:
+        async with ddb_client(profile=profile) as ddb:
+            # Filter unwritten items.
+            items = item_filter(items)
+            response = await ddb.batch_get_item(
+                RequestItems={table: {"Keys": items}}
+            )
+
+            # Attempt deletion of items present on table.
+            items = item_filter(response["Responses"][table])
+            if items:
+                response = await ddb.batch_write_item(
+                    RequestItems={
+                        table: [
+                            {"DeleteRequest": {"Key": item} for item in items}
+                        ]
+                    }
+                )
+    except Exception as exc:
+        LOG.exception(exc)
+        raise exc
```

Comment:
The logging method here is not correct I believe, as the point of the `LOG.exception` method is that it already implicitly logs the current exception, i.e. you're meant to use it like:

```
LOG.exception("Error while doing foo")
```

...and then it'll result in a log message like:

```
Error while doing foo
<full exception details>
```

If you pass the exception to `LOG.exception` then I think you're going to just get the exception logged twice, e.g.

```
<str(exc)>
<full exception details>
```

You also don't have to explicitly catch the exception object because LOG.exception knows how to access it already.  So, in summary I think this code should be more like:

```suggestion
    except Exception:
        LOG.exception("Exception while rolling back %s items on %s", len(items), table)
        raise
```